bootstrap/secret/secure.go

```diff
@@ -0,0 +1,180 @@
+func (p *SecureProvider) SetClient(client secrets.SecretClient) {
+	p.secretClient = client
+}
+
+// GetSecrets retrieves secrets from a secret store.
+// path specifies the type or location of the secrets to retrieve.
+// keys specifies the secrets which to retrieve. If no keys are provided then all the keys associated with the
+// specified path will be returned.
+func (p *SecureProvider) GetSecrets(path string, keys ...string) (map[string]string, error) {
+	if cachedSecrets := p.getSecretsCache(path, keys...); cachedSecrets != nil {
+		return cachedSecrets, nil
+	}
+
+	if p.secretClient == nil {
+		return nil, errors.New("can't get secrets. Secure secret provider is not properly initialized")
+	}
+
+	secureSecrets, err := p.secretClient.GetSecrets(path, keys...)
+	if err != nil {
+		return nil, err
+	}
+
+	p.updateSecretsCache(path, secureSecrets)
+	return secureSecrets, nil
+}
+
+func (p *SecureProvider) getSecretsCache(path string, keys ...string) map[string]string {
+	secureSecrets := make(map[string]string)
+
+	// Synchronize cache access
+	p.cacheMutex.Lock()
+	defer p.cacheMutex.Unlock()
+
+	// check cache for keys
+	allKeysExistInCache := false
+	cachedSecrets, cacheExists := p.secretsCache[path]
+	value := ""
+
+	if cacheExists {
+		for _, key := range keys {
+			value, allKeysExistInCache = cachedSecrets[key]
+			if !allKeysExistInCache {
+				return nil
+			}
+			secureSecrets[key] = value
+		}
+
+		// return secureSecrets if the requested keys exist in cache
+		if allKeysExistInCache {
+			return secureSecrets
+		}
+	}
+
+	return nil
+}
+
+func (p *SecureProvider) updateSecretsCache(path string, secrets map[string]string) {
+	// Synchronize cache access
+	p.cacheMutex.Lock()
+	defer p.cacheMutex.Unlock()
+
+	if _, cacheExists := p.secretsCache[path]; !cacheExists {
+		p.secretsCache[path] = secrets
+	}
+
+	for key, value := range secrets {
+		p.secretsCache[path][key] = value
+	}
+}
+
+// StoreSecrets stores the secrets to a secret store.
+// it sets the values requested at provided keys
+// path specifies the type or location of the secrets to store
+// secrets map specifies the "key": "value" pairs of secrets to store
+func (p *SecureProvider) StoreSecrets(path string, secrets map[string]string) error {
+	if p.secretClient == nil {
+		return errors.New("can't store secrets. Secure secret provider is not properly initialized")
+	}
+
+	err := p.secretClient.StoreSecrets(path, secrets)
+	if err != nil {
+		return err
+	}
+
+	// Synchronize cache access before clearing
+	p.cacheMutex.Lock()
+	// Clearing cache because adding a new secret(p) possibly invalidates the previous cache
+	p.secretsCache = make(map[string]map[string]string)
+	p.cacheMutex.Unlock()
+	//indicate to the SDK that the cache has been invalidated
+	p.lastUpdated = time.Now()
+	return nil
+}
+
+// SecretsUpdated is not need for secure secrets as this is handled when secrets are stored.
+func (p *SecureProvider) SecretsUpdated() {
+	// Do nothing
```

Comment:
should we consider moving `p.lastUpdated = time.Now()` to here, and call `p.SecretsUpdated()` in the StoreSecrets func?

Reply:
No, see comment about this is handled in StoreSecrets. This interface is here for Insecure Secrets and not used for the Secure secrets.